hw4/04-templates.js

```diff
@@ -3,62 +3,118 @@ const axios = require('axios');
 
 const app = express();
 const port = process.env.PORT || 5001;
+const url = 'https://restcountries.com/v3.1/all';
+
 
-// Use Pug as the templating engine
 app.set('views', __dirname + '/views');
 app.set('view engine', 'pug');
 
-// REST Countries URL
-const url = 'https://restcountries.com/v3.1/all';
 
-// Add your code here
+let groupBy = function(data, key) {
+  return data.reduce(function(storage, item) {
+    let group = item[key];
+
+    storage[group] = storage[group] || [];
+    storage[group] = Number(storage[group]) + item.population;
+
+    return storage; 
+  }, {});
+};
+
+
+const instructions = [
+  { 'title'     : '/capitals',
+    'params'    : 'capital',
+    'command'   : '`${country.name.official} - ${country.capital}`',
+  },
+  {
+    'title'     : '/populous',
+    'params'    : 'population',
+    'command'   : 'if (country.population >= 50000000) `${country.name.official} - ${country.population}`',
+  },
+  {
+    'title'     : '/regions',
+    'params'    : 'region',
+    'command'   : 'groupBy(response.data, "region")'
+  }
+];
+
+
+const extractData = (instruction) => {
+  
+  return async (req, res, next) => {
+    let dataToRender = [];
+
+    await axios.get(url).then((response) => {
+      if (response.data) {
+        if (instruction.title === '/regions') {
+          dataToRender.push(eval(instruction.command));
+        }
+        else {
+          response.data.forEach((country) => {
+            if (country.capital)
+              dataToRender.push(eval(instruction.command));
+          });
+        }
+      }
+    }).catch((error) => {
+      console.log(`The GET request to Axios (url: ${url}) failed; error: ${error}`);
+    });
+    
+    res.locals.dataToRender = dataToRender.filter((element) => element !== undefined).sort();
```

Comment:
/capitals
❏ Return the list of all the countries in the world and their
capitals in alphabetical order.

/populous
❏ Return the list of countries and their populations, starting with the most
populous.

In both cases sorting logic is different

Reply:
You're right! I forgot to sort by population there. Thank you.